### PR TITLE
Feature/naming convention

### PR DESCRIPTION
**Pull request by joaandrade**
Fixes #

### Proposed fixes:
To use the snakecase naming convention in oder to standardize all the metadata field names in the NextGEOSS Catalogue.
e.g. Point of Contact -> point_of_contact

### Features:
Function **convert_to_clean_snakecase**  added to the nextgeoss_base.py. This function converts all the extra field keys to the snacase format.

- [X] includes new  features
- [ ] includes tests covering changes
- [ ] includes updated documentation
- [ ] includes user-visible changes
- [ ] includes bugfix

Please [X] all the boxes above that apply

### CODE REVIEW

**Comment by joaandrade** [comment]
@georgiana-b @JGulic, we tested locally the naming convention function for two different situations:
- Harvest new products
- Update old harvested products (update_all flag)

Please check if everything is working properly.
Hope this helps.
João

File: ckanext/nextgeossharvest/lib/nextgeoss_base.py [hunk 1]
```diff
@@ -424,6 +427,10 @@ def _crawl_urls_ftp(self, url, provider):
                                                      status_code, elapsed))
         return size
 
+    def convert_to_clean_snakecase(self, extra_key):
+        clean_extra_key = re.sub('[^0-9a-zA-Z]+', '_', stringcase.snakecase(extra_key)).strip('_')
```
Comment (georgiana-b) [inline ckanext/nextgeossharvest/lib/nextgeoss_base.py:431]: @joaandrade I wonder why was this replace needed: `re.sub('[^0-9a-zA-Z]+', '_'`. 
Could you explain?

Comment (joaandrade) [reply ckanext/nextgeossharvest/lib/nextgeoss_base.py:431]: Hi @georgiana-b ,
We performed tests for the EBAS NILU datasets which are the one with the most incoherent naming convention. Find below some examples where the re.sub() and the strip() are necessary. The snakecase will replace every non alphanumeric chars (and spaces too) by a "_".  Also uppercase letters are replaced. This way, sometimes we will have multiple "_". With this line of code we are replacing all sequences of special chars by a single "_" and removing if it is at the beginning or at the end of the string.

snakecase_eg_1
snakecase_eg_2

Comment (joaandrade) [reply ckanext/nextgeossharvest/lib/nextgeoss_base.py:431]: @georgiana-b does this answer your question?

**Comment by joaandrade** [comment]
IMPORTANT NOTE: By accepting this naming convention solution, some metadata fields considered in the ignore_list code won't be ignored anymore:
https://github.com/NextGeoss/ckanext-nextgeossharvest/tree/master/ckanext/nextgeossharvest/ignore_list

**Comment by georgiana-b** [comment]
@joaandrade I added a unit test just to make sure we're successfully covering all the cases in the issue.

You're right about the list of ignored fields. I personally think those changes should be included in this pull request since they basically belong to this feature.

We will also have some pull requests on our side to rename some variables that we used in the code and when we're done  we will merge all of them simultaneously.

Could you please make sure you include any other renaming that you need to do in your code in this PR?

**Comment by joaandrade** [comment]
Thank you for the unit test @georgiana-b . We will implement a way to fix the ignore_list and use this pull request to commit the changes.

**Comment by joaandrade** [comment]
@georgiana-b after a discussion with @davidcordeiro-deimos we figured out that this will affect more than the ignore lists. Other extra metadata fields already harvested, that we want to keep in SOLR, but with an updated key, will also be affected. This means that this task is not that straightforward.
I will add this task to our backlog and we will always use this pull request and branch for these implementations.

**Comment by georgiana-b** [comment]
I also tried to run these changes today and I'm getting this error on indexing, both for existing datasets and for new ones:

```
[ckanext.nextgeossharvest.lib.nextgeoss_base.import] Package 85e6c91d-f90f-4031-9bb1-c77a4ae976d9 was successully harvested.
ng_ckan_dev   | 2020-04-14 14:00:10,321 INFO  [ckanext.harvest.queue] Received harvest object id: ed9b4773-aaaa-4241-8510-ff1ed58eaeb5
ng_ckan_dev   | 2020-04-14 14:00:10,343 DEBUG [ckanext.nextgeossharvest.lib.nextgeoss_base.import] Import stage for harvest object with GUID ed9b4773-aaaa-4241-8510-ff1ed58eaeb5
ng_ckan_dev   | 2020-04-14 14:00:10,344 DEBUG [ckanext.nextgeossharvest.lib.nextgeoss_base] Using config: {u'end_date': u'2019-01-01T12:00:00.000Z', u'update_all': False, u'make_private': False, u'source': u'esa_scihub', u'skip_raw': True, u'timeout': 100, u'datasets_per_job': 99, u'start_date': u'2019-01-01T11:00:00.000Z'}
ng_ckan_dev   | 2020-04-14 14:00:10,361 DEBUG [ckanext.nextgeossharvest.lib.nextgeoss_base] Updating s2b_msil1c_20190101t073319_n0207_r049_t37mcs_20190101t093045
ng_solr       | 503190 INFO  (qtp1989972246-17) [   x:ckan] o.a.s.c.S.Request [ckan]  webapp=/solr path=/select params={q=name:"s2b_msil1c_20190101t073319_n0207_r049_t37mcs_20190101t093045"+OR+id:"s2b_msil1c_20190101t073319_n0207_r049_t37mcs_20190101t093045"&fq=site_id:"default"&rows=1&wt=json} hits=0 status=0 QTime=0
ng_ckan_dev   | 2020-04-14 14:00:10,548 ERROR [ckan.lib.search] can't compare offset-naive and offset-aware datetimes
ng_ckan_dev   | Traceback (most recent call last):
ng_ckan_dev   |   File "/srv/app/src/ckan/ckan/lib/search/__init__.py", line 97, in dispatch_by_operation
ng_ckan_dev   |     index.update_dict(entity)
ng_ckan_dev   |   File "/srv/app/src/ckan/ckan/lib/search/index.py", line 101, in update_dict
ng_ckan_dev   |     self.index_package(pkg_dict, defer_commit)
ng_ckan_dev   |   File "/srv/app/src/ckan/ckan/lib/search/index.py", line 238, in index_package
ng_ckan_dev   |     if date != bogus_date:
ng_ckan_dev   | TypeError: can't compare offset-naive and offset-aware datetimes
ng_ckan_dev   | 2020-04-14 14:00:10,549 ERROR [ckan.model.modification] can't compare offset-naive and offset-aware datetimes
ng_ckan_dev   | Traceback (most recent call last):
ng_ckan_dev   |   File "/srv/app/src/ckan/ckan/model/modification.py", line 88, in notify
ng_ckan_dev   |     observer.notify(entity, operation)
ng_ckan_dev   |   File "/srv/app/src/ckan/ckan/lib/search/__init__.py", line 123, in notify
ng_ckan_dev   |     operation
ng_ckan_dev   |   File "/srv/app/src/ckan/ckan/lib/search/__init__.py", line 97, in dispatch_by_operation
ng_ckan_dev   |     index.update_dict(entity)
ng_ckan_dev   |   File "/srv/app/src/ckan/ckan/lib/search/index.py", line 101, in update_dict
ng_ckan_dev   |     self.index_package(pkg_dict, defer_commit)
ng_ckan_dev   |   File "/srv/app/src/ckan/ckan/lib/search/index.py", line 238, in index_package
ng_ckan_dev   |     if date != bogus_date:
ng_ckan_dev   | TypeError: can't compare offset-naive and offset-aware datetimes
ng_solr       | 503385 INFO  (qtp1989972246-12) [   x:ckan] o.a.s.c.S.Request [ckan]  webapp=/solr path=/select params={q=name:"6e3aaa0c-f458-4a4d-917d-4c39908df5dd"+OR+id:"6e3aaa0c-f458-4a4d-917d-4c39908df5dd"&fq=site_id:"default"&rows=1&wt=json} hits=0 status=0 QTime=0
```

Can you reproduce this as well? It could have something to do with the fact that I merged master into this branch so please give it a try again. 
I got this error on both Sentinel and EBAS harvesters.

**Comment by joaandrade** [comment]
@georgiana-b is it possible to unmerge? We tried this branch before without errors. However, we know that the naming covention will bring other issues that shall be addressed before (ignore list, etc.). I suggest:
To undo the merge of the master into this branch and then we will address all the issues that may appear with the naming convention. Let's put this PR onhold.

**Comment by georgiana-b** [comment]
I merged because I needed the fix to the plugin in order run the tests.

But ok, just to avoid confusion for you I reverted the merge. 

Just be aware that we will have had to merge master into this branch eventually so if the error does occur we will need to address it anyway. 

As for putting it on hold, I understand you have other priorities now so feel free to address it in your own time.

**Comment by joaandrade** [comment]
Thank you @georgiana-b. We have now an idea about what can be the issues to address. We will start this activity as soon as possible.

**Comment by JGulic** [comment]
Just to note, that this PR introduces changes in other places as well. For now it will stay as an open PR

**Comment by joaandrade** [comment]
This is an issue we will analyse later.

**Comment by davidcordeiro-deimos** [comment]
@georgiana-b a new commit as been performed to solve the issue that you experienced previously.

Just to explain briefly the error that you found. It was occurring during the package dictionary was being sent to the SOLR. If you check the CKAN source code [here](https://github.com/ckan/ckan/blob/534fd5088d6b8140b332bbc1e4280769143d404a/ckan/lib/search/index.py#L245), you can see that it parses fields in the dictionary that end with "_date". However, this source code is not able to handle dates that end with "Z. Thus, reporting the error that you got.

This is an error that could have occurred in any harvester as soon as we started to index each relevant field that ended with "_date". You were able to find this issue in Sentinel and EPOS harvesters because they had the fields "IngestionDate" and "ProcessingDate", respectively. Which, after the snake case transformation would become "ingestion_date" and "processing_date".

The easiest solution that I found was to simply remove "Z" from all the fields that end with "_date" in the nextgeoss_base.py.

In the scope of this PR, I also made sure that all comparisons with the ignore_list now occur with the name convention, even if the fields in the Ignore List do not follow it.

If you have any further comment please let me know.

**Comment by davidcordeiro-deimos** [comment]
@JGulic @georgiana-b during the review of this Pull Request, please take into consideration the impact of this PR upon the OpenSearch [repository](https://github.com/NextGeoss/ckanext-opensearch)

**Comment by joaandrade** [comment]
@georgiana-b , @JGulic 
We confirmed that the update in the naming convention will affect the OpenSearch, mainly for Sentinel parameters added such as orbit direction, cloud coverage, etc.

For example in the sentinel datasets, we are storing the orbit direction as **OrbitDirection**. And the Opensearch is expecting this. Applying the snakecase naming convention, the parameter will be stored as **orbit_direction** and the Opensearch response will contain empty fields: <eo:orbitDirection></eo:orbitDirection>

I confirmed and the datasets have this field with a value (DESCENDING or ASCENDING). But after applying the snakecase, the values does not appear anymore in the Opensearch response.

So in conclusion, the naming convention will affect the Opensearch.
